data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/resync/CascadingActionDetector.java

```diff
@@ -0,0 +1,187 @@
+        final Map<String, ParentTable> parentTableMap = new HashMap<>();
+        if (streamPartition.getProgressState().isEmpty()) {
+            return parentTableMap;
+        }
+
+        List<ForeignKeyRelation> foreignKeyRelations = streamPartition.getProgressState().get().getForeignKeyRelations();;
+
+        for (ForeignKeyRelation foreignKeyRelation : foreignKeyRelations) {
+            if (!ForeignKeyRelation.containsCascadingAction(foreignKeyRelation)) {
+                // skip foreign key relations without cascading actions
+                continue;
+            }
+
+            final String fullParentTableName = getFullTableName(foreignKeyRelation.getDatabaseName(), foreignKeyRelation.getParentTableName());
+            ParentTable parentTable;
+            if (!parentTableMap.containsKey(fullParentTableName)) {
+                Map<String, List<ForeignKeyRelation>> referencedColumnMap = new HashMap<>();
+                referencedColumnMap.put(foreignKeyRelation.getReferencedKeyName(), new ArrayList<>(List.of(foreignKeyRelation)));
+                parentTable = ParentTable.builder()
+                        .databaseName(foreignKeyRelation.getDatabaseName())
+                        .tableName(foreignKeyRelation.getParentTableName())
+                        .referencedColumnMap(referencedColumnMap)
+                        .build();
+                parentTableMap.put(fullParentTableName, parentTable);
+            } else {
+                parentTable = parentTableMap.get(fullParentTableName);
+                if (!parentTable.getReferencedColumnMap().containsKey(foreignKeyRelation.getReferencedKeyName())) {
+                    parentTable.getReferencedColumnMap().put(foreignKeyRelation.getReferencedKeyName(), new ArrayList<>());
+                }
+                parentTable.getReferencedColumnMap().get(foreignKeyRelation.getReferencedKeyName()).add(foreignKeyRelation);
+            }
+        }
+        LOG.debug("ParentTables are {}", parentTableMap.keySet());
+        return parentTableMap;
+    }
+
+    /**
+     * Detects if a binlog event contains cascading updates and if detected, creates resync partitions
+     */
+    public void detectCascadingUpdates(Event event, Map<String, ParentTable> parentTableMap, TableMetadata tableMetadata) {
+        final UpdateRowsEventData data = event.getData();
+        if (parentTableMap.containsKey(tableMetadata.getFullTableName())) {
+            final ParentTable parentTable = parentTableMap.get(tableMetadata.getFullTableName());
+
+            for (Map.Entry<Serializable[], Serializable[]> row : data.getRows()) {
+                // Find out for this row, which columns are changing
+                LOG.debug("Checking for updated columns");
+                final Map<String, Object> updatedColumnsAndValues = IntStream.range(0, row.getKey().length)
+                        .filter(i -> !row.getKey()[i].equals(row.getValue()[i]))
```

Comment:
what does row.getKey() and row.getValue() refer to ? Are these column names of changes columns ?